docs/standard/native-interop/best-practices.md

```diff
@@ -182,9 +182,9 @@ The following types are the same size on 32-bit and 64-bit Windows, despite thei
 | 16    | `WORD`           | `unsigned short`     | `ushort` |                                      |
 | 16    | `ATOM`           | `unsigned short`     | `ushort` |                                      |
 | 32    | `INT`            | `int`                | `int`    |                                      |
-| 32    | `LONG`           | `long`               | `int`    |                                      |
-| 32    | `ULONG`          | `unsigned long`      | `uint`   |                                      |
-| 32    | `DWORD`          | `unsigned long`      | `uint`   |                                      |
+| 32    | `LONG`           | `long`               | `int`    |  See `CLong` and `CULong` below.     |
+| 32    | `ULONG`          | `unsigned long`      | `uint`   |  See `CLong` and `CULong` below.     |
+| 32    | `DWORD`          | `unsigned long`      | `uint`   |  See `CLong` and `CULong` below.     |
```

Comment:
What about `HRESULT` and `NTSTATUS` below? They are also Windows `long` and therefore on Windows only should be `int`/`uint`.

However, if that code is ported one-to-one to Windows (like was essentially done for https://devblogs.microsoft.com/directx/directx-heart-linux/, https://github.com/microsoft/DirectX-Headers/blob/main/include/wsl), they start requiring `CLong` and `CULong`.

Reply:
They actually did the right thing there. This should really be `unsigned int`. A 64-bit `HRESULT` is going to be a nightmare to convert.

https://github.com/microsoft/DirectX-Headers/blob/5b5b57aee33e8e38eff0242221893a43ac6ea90d/include/wsl/winadapter.h#L24

https://github.com/microsoft/DirectX-Headers/blob/5b5b57aee33e8e38eff0242221893a43ac6ea90d/include/wsl/winadapter.h#L185

Reply:
That depends on how you choose to build your Windows emulator. NTSTATUS is 32-bit in CoreCLR PAL that number of projects copied Windows emulator layer from.

Reply:
I don't think the change of `HRESULT`/`NTSTATUS` in the table to `int` is correct. They are fundamentally the `long` type in C and on Unix systems, without a typedef fixup, will be 32-64 bits.

Yes, the Windows intent is that it is 32-bits, same as the Windows intent is that DWORD is 32-bits. But, in practice it only matters what is used to compile on Unix which may still be `long` and which may not be retyped.

It is 100% a nightmare and it likewise causes issues between `Unix` and `Windows` because on `Windows` you can't represent a full point but the Unix intent might have been that it is a pointer.

I did miss that WSL had `typedef'd` `LONG` as `int32_t`; ut they've not handled other cases. `HWND` for example is a proper `void*` on Windows, but an `int` in the WSL header. `wchar_t` is a `uint16_t` on Windows and a `uint32_t` on Unix. `LPARAM; and `LONG_PTR` is a pointer on Windows, while they've typedef'd to `int64_t` 😄

Reply:
Yes, building Windows emulation layers is fun and everybody does it slightly differently dependending on what makes the most sense for their specific component.